rest-api/spring-rest/src/main/java/uk/gov/gchq/gaffer/rest/controller/GremlinController.java

```diff
@@ -0,0 +1,188 @@
+import uk.gov.gchq.gaffer.operation.OperationChain;
+import uk.gov.gchq.gaffer.rest.factory.spring.AbstractUserFactory;
+import uk.gov.gchq.gaffer.tinkerpop.GafferPopGraph;
+import uk.gov.gchq.gaffer.tinkerpop.GafferPopGraphVariables;
+
+import java.nio.charset.StandardCharsets;
+import java.util.LinkedList;
+
+import static org.springframework.http.MediaType.APPLICATION_JSON_VALUE;
+import static org.springframework.http.MediaType.TEXT_PLAIN_VALUE;
+
+@RestController
+@Tag(name = "gremlin")
+@RequestMapping("/rest/gremlin")
+public class GremlinController {
+
+    // Keys for response JSON
+    public static final String EXPLAIN_OVERVIEW_KEY = "overview";
+    public static final String EXPLAIN_OP_CHAIN_KEY = "chain";
+    public static final String EXPLAIN_GREMLIN_KEY = "gremlin";
+
+    private final ConcurrentBindings bindings = new ConcurrentBindings();
+    private final AbstractUserFactory userFactory;
+    private final Graph graph;
+
+    @Autowired
+    public GremlinController(final GraphTraversalSource g, final AbstractUserFactory userFactory) {
+        bindings.putIfAbsent("g", g);
+        graph = g.getGraph();
+        this.userFactory = userFactory;
+    }
+
+    /**
+     * Explains what Gaffer operations are ran for a given gremlin query.
+     *
+     * @param httpHeaders The request headers.
+     * @param gremlinQuery The gremlin groovy query.
+     * @return Json response with explanation in.
+     */
+    @PostMapping(path = "/explain", consumes = TEXT_PLAIN_VALUE, produces = APPLICATION_JSON_VALUE)
+    @io.swagger.v3.oas.annotations.Operation(
+        summary = "Explain a Gremlin Query",
+        description = "Runs a Gremlin query and outputs an explanation of what Gaffer operations were executed on the graph")
+    public String explain(@RequestHeader final HttpHeaders httpHeaders, @RequestBody final String gremlinQuery) {
+        return runGremlinAndGetExplain(gremlinQuery, httpHeaders).toString();
+    }
+
+    /**
+     * Explains what Gaffer operations are ran for a given cypher query,
+     * will translate to Gremlin using {@link CypherAst} before executing.
+     *
+     * @param httpHeaders The request headers.
+     * @param cypherQuery Opencypher query.
+     * @return Json response with explanation in.
+     */
+    @PostMapping(path = "/cypher/explain", consumes = TEXT_PLAIN_VALUE, produces = APPLICATION_JSON_VALUE)
+    @io.swagger.v3.oas.annotations.Operation(
+        summary = "Explain a Cypher Query Executed via Gremlin",
+        description = "Translates a Cypher query to Gremlin and outputs an explanation of what Gaffer operations were executed on the graph")
+    public String cypherExplain(@RequestHeader final HttpHeaders httpHeaders, @RequestBody final String cypherQuery) {
+
+        final CypherAst ast = CypherAst.parse(cypherQuery);
+        // Translate the cypher to gremlin, always add a .toList() otherwise Gremlin wont execute it as its lazy
```

Comment:
Is it worth putting the toList warning in the swagger docs?

Reply:
Good point, will make sure its in the main docs too.